addons/grafana/Dockerfile.grafana_ppc64le

```diff
@@ -0,0 +1,19 @@
+# starts graphana with Prometheus Datasource pre configured
```

Comment:
I am not sure this will scale. 

Would it be possible to change the makefile to do a small sed on the Dockerfile, and replace
'from' with a custom version ?

This will allow future arm, mips, etc - and may also allow using a different, custom base

Reply:
Agreed. It should be possible to do a sed on the Dockerfile. Will make that change.

Reply:
I think the sed approach adds a lot more complexity than the approach I suggested (over in Dockerfile.servicegraph_debug_ppc64le) of using a ARG variable in the from statement for the Dockerfile.  I've got a more concrete example here:

https://github.com/istio/istio/compare/master...mattdelco:master16

In that change only the changes to the Dockerfiles and istio-docker.mk are relevant.  That change is for a different use case of trying to consume a local proxy on an opt-in basis.  This change seems to be headed in the direction of always having the Makefile specify the value for the ARG, and that case should work fine too (the default value for the ARG in the Dockerfile can be omitted).  The main downside I noticed in my approach is that the same DOCKER_RULE is used for all docker images, and "docker build" logs a messages if the value of a --build-arg wasn't used in a Dockerfile (there's probably ways to address this if it annoys people).

Reply:
> I think the sed approach adds a lot more complexity than the approach I suggested (over in Dockerfile.servicegraph_debug_ppc64le) of using a ARG variable in the from statement for the Dockerfile.

@mattdelco Well, the sed approach certainly adds more lines. I have added a commit which uses the sed approach. I am fine with taking the other approach (i.e., the ARG approach).

> The main downside I noticed in my approach is that the same DOCKER_RULE is used for all docker images, and "docker build" logs a messages if the value of a --build-arg wasn't used in a Dockerfile (there's probably ways to address this if it annoys people).

Well, if it annoys people, can't we make `DOCKER_OPT` "" (empty) [here](https://github.com/istio/istio/compare/master...mattdelco:master16) based on whether it should be used? Or else, we can have two separate `DOCKER_RULE`s, one with the arg and other without?

Reply:
I'm thinking that the customizing of FROM might increase over time for various reasons, partly because there's 3 you want to customize, and I had a different reason for another (though a recent change made mine use case morph into a different kind of problem), so more are conceivable.

Regardless, I've been thinking about ways to appropriately filter/prune the DOCKER_OPT for each docker image.  My first line of thinking involved using "$(filter" to pick out options that contained the target name (and then possibly use "$(subst" to prune off the part that contained the target name)), but now I'm thinking that a more straightforward option is to just use $(DOCKER_OPT$(suffix $@)) rather than a plain $(DOCKER_OPT) in DOCKER_RULE.  Then opts can be defined on a per-image basis, e.g.:

<pre>ifeq ($(MACH), ppc64le)
DOCKER_OPT.mixer_debug:=--build-arg MIXER_DEBUG_BASE=ubuntu:xenial
DOCKER_OPT.servicegraph_debug:=--build-arg SERVICEGRAPH_DEBUG_BASE=ubuntu:xenial
DOCKER_OPT.grafana:=--build-arg GRAFANA_BASE=ibmcom/grafana-ppc64le:4.4.3
else
DOCKER_OPT.mixer_debug:=--build-arg MIXER_DEBUG_BASE=gcr.io/istio-testing/ubuntu_xenial_debug:3f57ae2aceef79e4000fb07ec850bbf4bce811e6f81dc8cfd970e16cdf33e622
DOCKER_OPT.servicegraph_debug:=--build-arg SERVICEGRAPH_DEBUG_BASE=gcr.io/istio-testing/ubuntu_xenial_debug:3f57ae2aceef79e4000fb07ec850bbf4bce811e6f81dc8cfd970e16cdf33e622
DOCKER_OPT.grafana:=--build-arg GRAFANA_BASE=grafana/grafana:4.1.2
endif
</pre>
and then using mixer debug as an example its Dockerfile would have:
<pre>
ARG MIXER_DEBUG_BASE
FROM ${MIXER_DEBUG_BASE}
</pre>
As a test of the approach I made a standalone Makefile like this:
<pre>
HUB:=hub
TAG:=tag
SECURITY_DOCKER:=docker.istio-ca docker.istio-ca-test docker.node-agent docker.node-agent-test docker.multicluster-ca

default: $(SECURITY_DOCKER)

DOCKER_OPT.istio-ca:=--build-arg OPT_CA=ca:latest
DOCKER_OPT.istio-ca-test:=--build-arg OPT_CA_TEST=catest:latest
DOCKER_OPT.node-agent:=--build-arg OPT_NODE_AGENT=nodeagent:latest
DOCKER_OPT.node-agent-test:=--build-arg OPT_NODE_AGENT_TEST=nodeagenttest:latest

$(SECURITY_DOCKER):
	$(DOCKER_RULE)

DOCKER_RULE=@echo "docker build $(DOCKER_OPT$(suffix $@)) -t $(HUB)/$(subst docker.,,$@):$(TAG) -f Dockerfile$(suffix $@) ."
</pre>
and if I ran "make default" it output the commands with the desired customization:

<pre>
docker build --build-arg OPT_CA=ca:latest -t hub/istio-ca:tag -f Dockerfile.istio-ca .
docker build --build-arg OPT_CA_TEST=catest:latest -t hub/istio-ca-test:tag -f Dockerfile.istio-ca-test .
docker build --build-arg OPT_NODE_AGENT=nodeagent:latest -t hub/node-agent:tag -f Dockerfile.node-agent .
docker build --build-arg OPT_NODE_AGENT_TEST=nodeagenttest:latest -t hub/node-agent-test:tag -f Dockerfile.node-agent-test .
docker build -t hub/multicluster-ca:tag -f Dockerfile.multicluster-ca .
</pre>

I did simplify my initial example above, and I wouldn't be surprised if people desired that the image names be put on simpler standalone lines so that the update bots have less complex lines to parse, e.g.:

<pre>ifeq ($(MACH), ppc64le)
MIXER_DEBUG_BASE:=ubuntu:xenial
SERVICEGRAPH_DEBUG_BASE:=ubuntu:xenial
GRAFANA_BASE:=ibmcom/grafana-ppc64le:4.4.3
else
MIXER_DEBUG_BASE:=gcr.io/istio-testing/ubuntu_xenial_debug:3f57ae2aceef79e4000fb07ec850bbf4bce811e6f81dc8cfd970e16cdf33e622
SERVICEGRAPH_DEBUG_BASE:=gcr.io/istio-testing/ubuntu_xenial_debug:3f57ae2aceef79e4000fb07ec850bbf4bce811e6f81dc8cfd970e16cdf33e622
GRAFANA_BASE:=grafana/grafana:4.1.2
endif

DOCKER_OPT.mixer_debug:=--build-arg MIXER_DEBUG_BASE=$(MIXER_DEBUG_BASE)
DOCKER_OPT.servicegraph_debug:=--build-arg SERVICEGRAPH_DEBUG_BASE=$(SERVICEGRAPH_DEBUG_BASE)
DOCKER_OPT.grafana:=--build-arg GRAFANA_BASE=$(GRAFANA_BASE)
</pre>

Reply:
@mattdelco thanks for the explanation. It does seem a more simple approach comparatively. I have added a [commit](https://github.com/istio/istio/pull/3684/commits/b597827b6600a062b474788850f0817d6aa191fe) here which implements the same.